blockchain/consensus/narwhal_cns/src/exec.rs

```diff
@@ -0,0 +1,174 @@
+    chain_store: Arc<ChainStore<BS>>,
+    output_tx: Sender<NarwhalOutput>,
+    locked: AtomicBool,
+}
+
+impl<BS> NarwhalExecutionState<BS> {
+    pub fn new(chain_store: Arc<ChainStore<BS>>, output_tx: Sender<NarwhalOutput>) -> Self {
+        Self {
+            chain_store,
+            output_tx,
+            locked: AtomicBool::new(false),
+        }
+    }
+}
+
+impl ExecutionStateError for NarwhalConsensusError {
+    fn node_error(&self) -> bool {
+        // For now treat all errors as fatal.
+        true
+    }
+}
+
+impl<BS> ExecutionState for NarwhalExecutionState<BS> {
+    type Error = NarwhalConsensusError;
+
+    fn ask_consensus_write_lock(&self) -> bool {
+        self.locked
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            .is_ok()
+    }
+
+    fn release_consensus_write_lock(&self) {
+        self.locked.store(false, Ordering::Relaxed)
+    }
+}
+
+#[async_trait]
+impl<BS> BatchExecutionState for NarwhalExecutionState<BS>
+where
+    BS: BlockStore + Sync + Send + 'static,
+{
+    /// Load the tip of the chain and get the last certificate index from the `Ticket` in the header.
+    async fn load_next_certificate_index(&self) -> Result<SequenceNumber, Self::Error> {
+        match self.chain_store.heaviest_tipset().await {
+            None => Ok(0),
+            Some(tipset) => {
+                let index = ConsensusTransactionIndex::try_from(tipset.as_ref())?;
+                Ok(index.next_consensus_index())
+            }
+        }
+    }
+
+    async fn handle_consensus(
+        &self,
+        consensus_output: &ConsensusOutput,
+        transaction_batches: Vec<Vec<SerializedTransaction>>,
+    ) -> Result<(), Self::Error> {
+        // Append the transactions and the certificate to a pending queue that we can take values from
+        // when we have extended the local blockchain with a block. It is tempting to try to keep the
+        // last block we built here and just keep appending on top of it, or to append it directly to
+        // the chainstore. But that would bypass all the regular channels in `TipsetProcessor` and
+        // `ChainMuxer`, which are quite difficult to grasp, making this a risky option.
+
+        let output = NarwhalOutput {
+            consensus_output: consensus_output.clone(),
+            transaction_batches,
+        };
+        self.output_tx
+            .send(output)
+            .await
+            .map_err(|_| NarwhalConsensusError::Other("Cannot send output.".into()))?;
+        Ok(())
+    }
+}
+
+/// Support crash recovery by storing the index of processed transactions in the block tickets.
+/// Looking at the these numbers we can decide if we need to replay from the current, or the next consensus index.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct ConsensusTransactionIndex {
+    /// The consensus index the transactions in the block came from.
+    pub consensus_index: SequenceNumber,
```

Comment:
Is the `consensus_index` synced the same as the Narwhal rounds? And are we syncing the `consensus_index` with tipset epochs (i.e. the height of our chain)?

Reply:
I'm not sure what you mean by "synced". If you're asking whether the consensus index is the same as Narwhal rounds: I don't think so. Narwhal rounds happen for everyone, like, each validator can produce a certificate in each round (and they have to try as well otherwise the next round won't happen). I think the certificate index comes from Bullshark emitting the certificates in their final order. 

I don't know if there can be any gaps in them, or whether they are contiguous. The Sui code didn't expect continuity, just monotonicity, because it's never notified about empty blocks. 

So, I'm not using it for anything other than crash recovery: we have to tell Narwhal which certificate index is the next we want to be replayed to us. 

I can't sync it exactly with the epoch because a certificate might contain so many transactions that it has to be put into multiple Filecoin blocks. That's not implemented yet but in theory it's possible.